rcl/src/rcl/introspection.c

```diff
@@ -0,0 +1,383 @@
+      case rcl_interfaces__msg__ServiceEventType__RESPONSE_RECEIVED:
+        serialized_message_ts = introspection_utils->response_type_support;
+        break;
+      default:
+        RCL_SET_ERROR_MSG("Invalid event type");
+        return RCL_RET_ERROR;
+    }
+
+    ret = rmw_serialize(ros_response_request, serialized_message_ts, &serialized_message);
+    if (RMW_RET_OK != ret) {
+      RCL_SET_ERROR_MSG(rmw_get_error_string().str);
+      return RCL_RET_ERROR;
+    }
+    rosidl_runtime_c__octet__Sequence__init(
+      &msg.serialized_event, serialized_message.buffer_length);
+    memcpy(msg.serialized_event.data, serialized_message.buffer, serialized_message.buffer_length);
+  }
+
+  // populate info message
+  msg_info.event_type = event_type;
+  msg_info.sequence_number = sequence_number;
+  rosidl_runtime_c__String__assign(&msg_info.service_name, introspection_utils->service_name);
+  rosidl_runtime_c__String__assign(&msg_info.event_name, introspection_utils->service_type_name);
+
+  builtin_interfaces__msg__Time timestamp;
+  builtin_interfaces__msg__Time__init(&timestamp);
+
+  rcl_time_point_value_t now;
+  ret = rcl_clock_get_now(introspection_utils->clock, &now);
+  if (RMW_RET_OK != ret) {
+    RCL_SET_ERROR_MSG(rmw_get_error_string().str);
+    return RCL_RET_ERROR;
+  }
+  timestamp.sec = RCL_NS_TO_S(now);
+  timestamp.nanosec = now % (1000LL * 1000LL * 1000LL);
+  msg_info.stamp = timestamp;
+
+  unique_identifier_msgs__msg__UUID uuid_msg;
+  unique_identifier_msgs__msg__UUID__init(&uuid_msg);
+  memcpy(uuid_msg.uuid, uuid, 16 * sizeof(uint8_t));
+  msg_info.client_id = uuid_msg;
+
+  msg.info = msg_info;
+
+  // and publish it out!
+  ret = rcl_publish(introspection_utils->publisher, &msg, NULL);
+  if (RMW_RET_OK != ret) {
+    RCL_SET_ERROR_MSG(rmw_get_error_string().str);
+    return RCL_RET_ERROR;
+  }
+
+  // clean up
+  if (introspection_utils->_content_enabled) {
+    ret = rmw_serialized_message_fini(&serialized_message);
+    if (RMW_RET_OK != ret) {
+      RCL_SET_ERROR_MSG(rmw_get_error_string().str);
+      return RCL_RET_ERROR;
+    }
+  }
+
+  rcl_interfaces__msg__ServiceEvent__fini(&msg);
+  return RCL_RET_OK;
+}
+
+rcl_ret_t rcl_service_introspection_enable(
+  rcl_service_introspection_utils_t * introspection_utils, const rcl_node_t * node,
+  rcl_allocator_t * allocator)
+{
+  rcl_ret_t ret;
+
+  introspection_utils->publisher = allocator->allocate(sizeof(rcl_publisher_t), allocator->state);
+  *introspection_utils->publisher = rcl_get_zero_initialized_publisher();
+  const rosidl_message_type_support_t * service_event_typesupport =
+    ROSIDL_GET_MSG_TYPE_SUPPORT(rcl_interfaces, msg, ServiceEvent);
+
+  rcl_publisher_options_t publisher_options = rcl_publisher_get_default_options();
+  ret = rcl_publisher_init(
+    introspection_utils->publisher, node, service_event_typesupport,
+    introspection_utils->service_event_topic_name, &publisher_options);
```

Comment:
I don't think it's worth consolidating into a reusable function because it is only performing the pretty standard-but-verbose-in-rcl task creating a publisher. In prior implementations I checked if it were enabled/disabled already in the server/client-specific functions I created but it does make more sense to put this logic into `rcl_service_introspection_enable`/disable. 
This has been done in https://github.com/ros2/rcl/pull/997/commits/e1a2273929921002d255de2244898befecdc35d4